File: src/java/Core/org/hypertable/AsyncComm/CommHeader.java
/**
 * Copyright (C) 2008 Doug Judd (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

package org.hypertable.AsyncComm;

import java.nio.ByteBuffer;

import org.hypertable.Common.Checksum;
import org.hypertable.Common.Error;
import org.hypertable.Common.HypertableException;

public class CommHeader {

    public static byte VERSION = 1;

    public static byte FIXED_LENGTH = 38;

    public static short FLAGS_BIT_REQUEST          = (short)0x0001;
    public static short FLAGS_BIT_IGNORE_RESPONSE  = (short)0x0002;
    public static short FLAGS_BIT_URGENT           = (short)0x0004;
    public static short FLAGS_BIT_PAYLOAD_CHECKSUM = (short)0x8000;

    public static short FLAGS_MASK_REQUEST          = (short)0xFFFE;
    public static short FLAGS_MASK_IGNORE_RESPONSE  = (short)0xFFFD;
    public static short FLAGS_MASK_URGENT           = (short)0xFFFB;
    public static short FLAGS_MASK_PAYLOAD_CHECKSUM = (short)0x7FFF;

    public CommHeader() {
        version = VERSION;
        header_len = FIXED_LENGTH;
    }

    public CommHeader(long cmd, int timeout) {
        version        = VERSION;
        header_len     = FIXED_LENGTH;
        command        = cmd;
        timeout_ms = timeout;
    }

    public CommHeader(long cmd) {
        version    = VERSION;
        header_len = FIXED_LENGTH;
        command    = cmd;
    }

    public int fixed_length() { return FIXED_LENGTH; }

    public int encoded_length() { return FIXED_LENGTH; }

    public void encode(ByteBuffer buf) {
        int saved_position = buf.position();
        int checksum_position;

        buf.put(version);
        buf.put(header_len);
        buf.putShort(alignment);
        buf.putShort(flags);
        checksum_position = buf.position();
        buf.putInt(0);  // 0 for checksum computation purposes
        buf.putInt(id);
        buf.putInt(gid);
        buf.putInt(total_len);
        buf.putInt(timeout_ms);
        buf.putInt(payload_checksum);
        buf.putLong(command);

        int header_length = buf.position() - saved_position;
        byte [] header_buffer = new byte [ header_length ];
        buf.position(saved_position);
        buf.get(header_buffer, 0, header_length);
        header_checksum = Checksum.fletcher32(header_buffer, 0, header_length);
        buf.position(checksum_position);
        buf.putInt(header_checksum);
        buf.position(saved_position + header_length);
    }

    public void decode(ByteBuffer buf) throws HypertableException {
        int saved_position = buf.position();

        version = buf.get();
        header_len = buf.get();
        alignment = buf.getShort();
        flags = buf.getShort();
        buf.mark();
        header_checksum = buf.getInt();
        id = buf.getInt();
        gid = buf.getInt();
        total_len = buf.getInt();
        timeout_ms = buf.getInt();
        payload_checksum = buf.getInt();
        command = buf.getLong();

        int header_length = buf.position() - saved_position;
        buf.reset();
        buf.putInt(0);
        byte [] header_buffer = new byte [ header_length ];
        buf.position(saved_position);
        buf.get(header_buffer, 0, header_length);
        int computed_checksum = Checksum.fletcher32(header_buffer, 0,
                                                    header_length);
        if (computed_checksum != header_checksum)
            throw new HypertableException(Error.COMM_HEADER_CHECKSUM_MISMATCH);
        buf.position(saved_position + header_length);
    }

    public void set_total_length(int len) { total_len = len; }

    public void initialize_from_request_header(CommHeader req_header) {
      flags = req_header.flags;
      id = req_header.id;
      gid = req_header.gid;
      command = req_header.command;
      total_len = 0;
    }

    public byte version;
    public byte header_len;
    public short alignment;
    public short flags;
    public int header_checksum;
    public int id;
    public int gid;
    public int total_len;
    public int timeout_ms;
    public int payload_checksum;
    public long command;
}

File: src/cc/AsyncComm/CommHeader.cc
/**
 * Copyright (C) 2008 Doug Judd (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */
#include "Common/Compat.h"
#include "Common/Checksum.h"
#include "Common/Error.h"
#include "Common/Serialization.h"
#include "Common/Logger.h"

#include "CommHeader.h"

using namespace Hypertable;

void CommHeader::encode(uint8_t **bufp) {
  uint8_t *base = *bufp;
  Serialization::encode_i8(bufp, version);
  Serialization::encode_i8(bufp, header_len);
  Serialization::encode_i16(bufp, alignment);
  Serialization::encode_i16(bufp, flags);
  Serialization::encode_i32(bufp, 0);
  Serialization::encode_i32(bufp, id);
  Serialization::encode_i32(bufp, gid);
  Serialization::encode_i32(bufp, total_len);
  Serialization::encode_i32(bufp, timeout_ms);
  Serialization::encode_i32(bufp, payload_checksum);
  Serialization::encode_i64(bufp, command);
  // compute and serialize header checksum
  header_checksum = fletcher32(base, (*bufp)-base);
  base += 6;
  Serialization::encode_i32(&base, header_checksum);
}

void CommHeader::decode(const uint8_t **bufp, size_t *remainp) {
  const uint8_t *base = *bufp;
  if (*remainp < FIXED_LENGTH)
    HT_THROWF(Error::COMM_BAD_HEADER,
              "Header size %d is less than the minumum fixed length %d",
              (int)*remainp, (int)FIXED_LENGTH);
  HT_TRY("decoding comm header",
         version = Serialization::decode_i8(bufp, remainp);
         header_len = Serialization::decode_i8(bufp, remainp);
         alignment = Serialization::decode_i16(bufp, remainp);
         flags = Serialization::decode_i16(bufp, remainp);
         header_checksum = Serialization::decode_i32(bufp, remainp);
         id = Serialization::decode_i32(bufp, remainp);
         gid = Serialization::decode_i32(bufp, remainp);
         total_len = Serialization::decode_i32(bufp, remainp);
         timeout_ms = Serialization::decode_i32(bufp, remainp);
         payload_checksum = Serialization::decode_i32(bufp, remainp);
         command = Serialization::decode_i64(bufp, remainp));
  memset((void *)(base+6), 0, 4);
  uint32_t checksum = fletcher32(base, *bufp-base);
  if (checksum != header_checksum)
    HT_THROWF(Error::COMM_HEADER_CHECKSUM_MISMATCH, "%u != %u", checksum,
              header_checksum);
}
